{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc17cbc1-02ec-457d-a6e3-de4afea40164",
   "metadata": {},
   "source": [
    "# 외부 라이브러리"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66e7aac-c972-4cc2-9750-ae8381b546ac",
   "metadata": {},
   "source": [
    "## pip *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e67c5a04-54a1-4164-a781-b52d95b91362",
   "metadata": {},
   "source": [
    "- venv1 생성 후 백업: pip freeze > requirements.txt\n",
    "- venv2 생성 후 복구: pip install -r requirements.txt\n",
    "- venv3 jupyterlab 설치후 matplotlib 설치 후 백업하고\n",
    "- venv4에 pip 리스트 그대로 복구하라.\n",
    "- venv5에 numpy 2.2.2 버전으로 복구되도록 requirements.txt 수정 후 복구하라."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc18a02a-2fec-4fd6-8bd2-84eed3dbba0d",
   "metadata": {},
   "source": [
    "## Faker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0ffeec4a-41af-42a9-ac53-c991688f3933",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install Faker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4253827c-f208-4883-956e-05114f5b2dd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': '김민지',\n",
       "  'id': 'iyujin',\n",
       "  'phone': '016-521-0305',\n",
       "  'email': 'mgim@example.com'},\n",
       " {'name': '김준영',\n",
       "  'id': 'soneunyeong',\n",
       "  'phone': '054-120-5259',\n",
       "  'email': 'jeonghyiju@example.org'},\n",
       " {'name': '최옥순',\n",
       "  'id': 'ggim',\n",
       "  'phone': '042-837-6447',\n",
       "  'email': 'hongyeonggil@example.net'},\n",
       " {'name': '서진호',\n",
       "  'id': 'di',\n",
       "  'phone': '041-506-1389',\n",
       "  'email': 'sanghyeoni@example.net'},\n",
       " {'name': '권지은',\n",
       "  'id': 'eungyeongyun',\n",
       "  'phone': '064-754-8819',\n",
       "  'email': 'donghyeonbag@example.com'},\n",
       " {'name': '우민지',\n",
       "  'id': 'jongsu39',\n",
       "  'phone': '061-419-8751',\n",
       "  'email': 'ogim@example.net'},\n",
       " {'name': '김주원',\n",
       "  'id': 'seongho60',\n",
       "  'phone': '070-0447-9826',\n",
       "  'email': 'gimyejin@example.net'},\n",
       " {'name': '배민서',\n",
       "  'id': 'iyang',\n",
       "  'phone': '043-353-3643',\n",
       "  'email': 'jaehogo@example.org'},\n",
       " {'name': '박정수',\n",
       "  'id': 'sugja33',\n",
       "  'phone': '063-091-2836',\n",
       "  'email': 'rgim@example.org'},\n",
       " {'name': '김미영',\n",
       "  'id': 'seoyeon35',\n",
       "  'phone': '02-9347-4321',\n",
       "  'email': 'jeongsun62@example.net'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from faker import Faker\n",
    "fake = Faker('ko-KR')\n",
    "[{'name':fake.name(), 'id':fake.user_name(),'phone':fake.phone_number(),'email':fake.email()} for n in range(10)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "52b15611-a3bc-4ae1-8594-7562dbd70f10",
   "metadata": {},
   "source": [
    "Faker 활용하기\n",
    "\n",
    "<table style=\"float:left;width:400px\">\n",
    "<thead>\n",
    "<tr>\n",
    "<th>항목</th>\n",
    "<th>설명</th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr>\n",
    "<td>fake.name()</td>\n",
    "<td>이름</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.address()</td>\n",
    "<td>주소</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.postcode()</td>\n",
    "<td>우편 번호</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.country()</td>\n",
    "<td>국가명</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.company()</td>\n",
    "<td>회사명</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.job()</td>\n",
    "<td>직업명</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.phone_number()</td>\n",
    "<td>휴대전화 번호</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.email()</td>\n",
    "<td>이메일 주소</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.user_name()</td>\n",
    "<td>사용자명</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.pyint(min_value=0, max_value=100)</td>\n",
    "<td>0부터 100 사이의 임의의 숫자</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.ipv4_private()</td>\n",
    "<td>IP 주소</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.text()</td>\n",
    "<td>임의의 문장 (한글 임의의 문장은 <code>fake.catch_phrase()</code> 사용)</td>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>fake.color_name()</td>\n",
    "<td>색상명</td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fe48b951-8b21-4f20-9fc7-4e35c938612a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1위:REBEL HEART/IVE (아이브)\n",
      "2위:HOME SWEET HOME (feat. 태양, 대성)/G-DRAGON\n",
      "3위:나는 반딧불/황가람\n",
      "4위:Whiplash/aespa\n",
      "5위:APT./로제 (ROSÉ)\n",
      "6위:오늘만 I LOVE YOU/Bruno Mars\n",
      "7위:Drowning/BOYNEXTDOOR\n",
      "8위:ATTITUDE/WOODZ\n",
      "9위:toxic till the end/IVE (아이브)\n",
      "10위:Dash/로제 (ROSÉ)\n",
      "11위:POWER/PLAVE\n",
      "12위:HAPPY/G-DRAGON\n",
      "13위:내게 사랑이 뭐냐고 물어본다면/DAY6 (데이식스)\n",
      "14위:RIZZ/로이킴\n",
      "15위:DRIP/PLAVE\n",
      "16위:Chroma Drift/BABYMONSTER\n",
      "17위:Island/PLAVE\n",
      "18위:소나기/PLAVE\n",
      "19위:12:32 (A to T)/이클립스 (ECLIPSE)\n",
      "20위:너와의 모든 지금/PLAVE\n",
      "21위:모르시나요(PROD.로코베리)/재쓰비 (JAESSBEE)\n",
      "22위:Supernova/조째즈\n",
      "23위:천상연/aespa\n",
      "24위:Welcome to the Show/이창섭\n",
      "25위:UP (KARINA Solo)/DAY6 (데이식스)\n",
      "26위:한 페이지가 될 수 있게/aespa\n",
      "27위:내 이름 맑음/DAY6 (데이식스)\n",
      "28위:사랑은 늘 도망가/QWER\n",
      "29위:슬픈 초대장/임영웅\n",
      "30위:어떻게 이별까지 사랑하겠어, 널 사랑하는 거지/순순희(지환)\n",
      "31위:Love wins all/AKMU (악뮤)\n",
      "32위:예뻤어/아이유\n",
      "33위:고민중독/DAY6 (데이식스)\n",
      "34위:Mantra/QWER\n",
      "35위:How Sweet/제니 (JENNIE)\n",
      "36위:우리들의 블루스/NewJeans\n",
      "37위:너의 모든 순간/임영웅\n",
      "38위:Supersonic/성시경\n",
      "39위:첫 만남은 계획대로 되지 않아/프로미스나인\n",
      "40위:Die With A Smile/TWS (투어스)\n",
      "41위:온기/Lady Gaga\n",
      "42위:나는 아픈 건 딱 질색이니까/Bruno Mars\n",
      "43위:그대만 있다면 (여름날 우리 X 너드커넥션 (Nerd Connection))/임영웅\n",
      "44위:Home/(여자)아이들\n",
      "45위:I AM/너드커넥션 (Nerd Connection)\n",
      "46위:무제(無題) (Untitled, 2014)/임영웅\n",
      "47위:모래 알갱이/IVE (아이브)\n",
      "48위:비의 랩소디/G-DRAGON\n",
      "49위:Small girl (feat. 도경수(D.O.))/임영웅\n",
      "50위:모든 날, 모든 순간 (Every day, Every Moment)/임재현\n",
      "51위:미안해 미워해 사랑해/이영지\n",
      "52위:다시 만날 수 있을까/폴킴\n",
      "53위:에피소드/Crush\n",
      "54위:Magnetic/임영웅\n",
      "55위:헤어지자 말해요/이무진\n",
      "56위:이제 나만 믿어요/아일릿(ILLIT)\n",
      "57위:다정히 내 이름을 부르면/박재정\n",
      "58위:Ditto/임영웅\n",
      "59위:Hype Boy/경서예지\n",
      "60위:청혼하지 않을 이유를 못 찾았어/전건호\n",
      "61위:number one girl/NewJeans\n",
      "62위:청춘만화/NewJeans\n",
      "63위:인사/이무진\n",
      "64위:해야 (HEYA)/로제 (ROSÉ)\n",
      "65위:Armageddon/이무진\n",
      "66위:Do or Die/범진\n",
      "67위:주저하는 연인들을 위해/IVE (아이브)\n",
      "68위:사랑인가 봐/aespa\n",
      "69위:무지개/임영웅\n",
      "70위:살기 위해서/잔나비\n",
      "71위:Igloo/멜로망스\n",
      "72위:London Boy/임영웅\n",
      "73위:Sticky/순순희\n",
      "74위:Supernatural/KISS OF LIFE\n",
      "75위:Seven (feat. Latto) - Clean Ver./임영웅\n",
      "76위:클락션 (Klaxon)/KISS OF LIFE\n",
      "77위:Polaroid/NewJeans\n",
      "78위:미치게 그리워서/정국\n",
      "79위:취중고백/(여자)아이들\n",
      "80위:Boom Boom Bass/임영웅\n",
      "81위:인생찬가/황가람\n",
      "82위:눈이 오잖아(Feat.헤이즈)/김민석\n",
      "83위:Hug (포옹)/RIIZE\n",
      "84위:Love 119/임영웅\n",
      "85위:녹아내려요/이무진\n",
      "86위:사막에서 꽃을 피우듯/RIIZE\n",
      "87위:To. X/RIIZE\n",
      "88위:봄여름가을겨울 (Still Life)/DAY6 (데이식스)\n",
      "89위:SPOT! (feat. JENNIE)/우디 (Woody)\n",
      "90위:사건의 지평선/태연 (TAEYEON)\n",
      "91위:청바지/BIGBANG (빅뱅)\n",
      "92위:Bubble Gum/지코 (ZICO)\n",
      "93위:봄날/윤하 (YOUNHA)\n",
      "94위:첫 눈/부석순 (SEVENTEEN)\n",
      "95위:Dynamite/NewJeans\n",
      "96위:Congratulations/방탄소년단\n",
      "97위:연애편지/EXO\n",
      "98위:Drama/방탄소년단\n",
      "99위:보금자리/DAY6 (데이식스)\n",
      "100위:earthquake/임영웅\n",
      " "
     ]
    }
   ],
   "source": [
    "import requests as bs\n",
    "from bs4 import BeautifulSoup as bs\n",
    "url = \"https://www.melon.com/chart/index.htm\"\n",
    "headers = {'User-Agent':'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36 Edg/133.0.0.0'}\n",
    "web = req.get(url, headers = headers)\n",
    "soup = bs(web.content, 'html.parser')\n",
    "\n",
    "title = soup.select('.rank01')\n",
    "name = soup.select('.checkEllipsis a')\n",
    "str = ''\n",
    "for i, (t,n) in enumerate(zip(title, name),1):\n",
    "    str += f'{i}위:{t.text.strip()}/{n.text}\\n'\n",
    "print(str,end=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c5749a5-1cd6-4225-8c0b-41e45d9227f3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'requests' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mmelon\u001b[39;00m\n\u001b[0;32m      2\u001b[0m melon\u001b[38;5;241m.\u001b[39mmel()\n",
      "File \u001b[1;32mD:\\lyb\\code\\melon.py:5\u001b[0m\n\u001b[0;32m      3\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://www.melon.com/chart/index.htm\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m      4\u001b[0m headers \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUser-Agent\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36 Edg/133.0.0.0\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[1;32m----> 5\u001b[0m web \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241m.\u001b[39mget(url, headers \u001b[38;5;241m=\u001b[39m headers)\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mmel\u001b[39m():\n\u001b[0;32m      7\u001b[0m     soup \u001b[38;5;241m=\u001b[39m bs(web\u001b[38;5;241m.\u001b[39mcontent, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhtml.parser\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'requests' is not defined"
     ]
    }
   ],
   "source": [
    "import melon\n",
    "melon.mel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d4d012a1-cd55-443e-a1b5-d575eb8400ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting html5lib\n",
      "  Downloading html5lib-1.1-py2.py3-none-any.whl.metadata (16 kB)\n",
      "Requirement already satisfied: six>=1.9 in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from html5lib) (1.17.0)\n",
      "Requirement already satisfied: webencodings in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from html5lib) (0.5.1)\n",
      "Downloading html5lib-1.1-py2.py3-none-any.whl (112 kB)\n",
      "Installing collected packages: html5lib\n",
      "Successfully installed html5lib-1.1\n"
     ]
    }
   ],
   "source": [
    "!pip install html5lib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e91b032c-a4e8-4922-a057-cbe8fd814129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===============\n",
      "월 2025.02.17\n",
      "정식-5,500원\n",
      "강황밥\n",
      "김치콩나물국\n",
      "돈육바싹불고기\n",
      "모듬잡채\n",
      "멸치새우견과볶음\n",
      "상추겉절이\n",
      "포기김치\n",
      "\n",
      "===============\n",
      "===============\n",
      "화 2025.02.18\n",
      "정식-5,500원\n",
      "백미밥\n",
      "물만두국\n",
      "순살닭매콤조림\n",
      "쫄면야채무침\n",
      "파래무무침\n",
      "참나물겉절이\n",
      "포기김치\n",
      "\n",
      "===============\n",
      "===============\n",
      "수 2025.02.19\n",
      "정식-5,500원\n",
      "백미밥\n",
      "모듬어묵국\n",
      "함박스테이크/치즈/소스\n",
      "스파게티/소스\n",
      "치커리유자청샐러드\n",
      "콩나물무침\n",
      "포기김치\n",
      "\n",
      "===============\n",
      "===============\n",
      "목 2025.02.20\n",
      "정식-5,500원\n",
      "잡곡밥\n",
      "홍합미역국\n",
      "순살치킨/맛초킹소스\n",
      "꼬시레기두부무침\n",
      "잡채어묵조림\n",
      "상추겉절이\n",
      "포기김치\n",
      "\n",
      "===============\n",
      "===============\n",
      "금 2025.02.21\n",
      "정식-5,500원\n",
      "백미밥\n",
      "국수\n",
      "깐풍기/소스\n",
      "알감자버터구이\n",
      "도토리묵김치무침\n",
      "열무겉절이\n",
      "포기김치\n",
      "\n",
      "===============\n"
     ]
    }
   ],
   "source": [
    "import requests as req\n",
    "from bs4 import BeautifulSoup as bs\n",
    "url = \"https://www.pusan.ac.kr/kor/CMS/MenuMgr/menuListOnBuilding.do?mCode=MN202\"\n",
    "headers = {'user-agent':'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36 Edg/133.0.0.0'}\n",
    "web = req.get(url, headers = headers)\n",
    "soup = bs(web.content, 'html5lib')\n",
    "menucard = soup.select('.menu-tb1')\n",
    "won = soup.select('.menu-tit01')\n",
    "menu = soup.select('h3.menu-tit01+p')\n",
    "day = soup.select('.day')\n",
    "date = soup.select('.date')\n",
    "for d, dd, w, m in zip(day, date, won, menu):\n",
    "    print('='*15)\n",
    "    print(d.text, dd.text)\n",
    "    print(w.text+'\\n'+m.text)\n",
    "    print('='*15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "093bc98c-1332-4c9e-ba24-46f32738d1d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting python-telegram-bot\n",
      "  Downloading python_telegram_bot-21.10-py3-none-any.whl.metadata (17 kB)\n",
      "Requirement already satisfied: httpx~=0.27 in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from python-telegram-bot) (0.28.1)\n",
      "Requirement already satisfied: anyio in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from httpx~=0.27->python-telegram-bot) (4.8.0)\n",
      "Requirement already satisfied: certifi in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from httpx~=0.27->python-telegram-bot) (2024.12.14)\n",
      "Requirement already satisfied: httpcore==1.* in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from httpx~=0.27->python-telegram-bot) (1.0.7)\n",
      "Requirement already satisfied: idna in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from httpx~=0.27->python-telegram-bot) (3.10)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from httpcore==1.*->httpx~=0.27->python-telegram-bot) (0.14.0)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\admin\\appdata\\local\\programs\\python\\python313\\lib\\site-packages (from anyio->httpx~=0.27->python-telegram-bot) (1.3.1)\n",
      "Downloading python_telegram_bot-21.10-py3-none-any.whl (669 kB)\n",
      "   ---------------------------------------- 0.0/669.5 kB ? eta -:--:--\n",
      "   --------------------------------------- 669.5/669.5 kB 31.0 MB/s eta 0:00:00\n",
      "Installing collected packages: python-telegram-bot\n",
      "Successfully installed python-telegram-bot-21.10\n"
     ]
    }
   ],
   "source": [
    "! pip install python-telegram-bot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b1c24d0-3592-4c5e-b975-3c0ebf5f7d28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting asyncio\n",
      "  Downloading asyncio-3.4.3-py3-none-any.whl.metadata (1.7 kB)\n",
      "Downloading asyncio-3.4.3-py3-none-any.whl (101 kB)\n",
      "Installing collected packages: asyncio\n",
      "Successfully installed asyncio-3.4.3\n"
     ]
    }
   ],
   "source": [
    "!pip install asyncio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff349071-8cbb-41b0-951d-facab7271022",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "봇이 실행중 입니다.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n",
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n",
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n",
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n",
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n",
      "Exception happened while polling for updates.\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 750, in _network_loop_retry\n",
      "    if not await do_action():\n",
      "           ^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 744, in do_action\n",
      "    return action_cb_task.result()\n",
      "           ~~~~~~~~~~~~~~~~~~~~~^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\futures.py\", line 199, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\asyncio\\tasks.py\", line 304, in __step_run_and_handle_result\n",
      "    result = coro.send(None)\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_updater.py\", line 371, in polling_action_cb\n",
      "    updates = await self.bot.get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 650, in get_updates\n",
      "    updates = await super().get_updates(\n",
      "              ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<9 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 4480, in get_updates\n",
      "    await self._post(\n",
      "    ^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    ),\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 619, in _post\n",
      "    return await self._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\ext\\_extbot.py\", line 354, in _do_post\n",
      "    return await super()._do_post(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\_bot.py\", line 648, in _do_post\n",
      "    result = await request.post(\n",
      "             ^^^^^^^^^^^^^^^^^^^\n",
      "    ...<6 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 202, in post\n",
      "    result = await self._request_wrapper(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    ...<7 lines>...\n",
      "    )\n",
      "    ^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python313\\Lib\\site-packages\\telegram\\request\\_baserequest.py\", line 385, in _request_wrapper\n",
      "    raise Conflict(message)\n",
      "telegram.error.Conflict: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running\n"
     ]
    }
   ],
   "source": [
    "import asyncio\n",
    "import nest_asyncio # 주피터랩용\n",
    "from dotenv import load_dotenv\n",
    "import os \n",
    "# .env 파일 로드\n",
    "load_dotenv()\n",
    "\n",
    "from telegram import Update\n",
    "from telegram.ext import Application, CommandHandler, MessageHandler, filters, CallbackContext\n",
    "nest_asyncio.apply() # 주피터랩용\n",
    "\n",
    "TELEGRAM_TOKEN = os.getenv(\"TELEGRAM_TOKEN\")\n",
    "\n",
    "async def start( update, context):\n",
    "    await context.bot.send_message(chat_id=update.effective_chat.id, text=\"봇이 가동되었습니다!\")\n",
    "async def handle_message(update , context ):\n",
    "    user_message = update.message.text.lower()  # Text -> text\n",
    "\n",
    "    if \"밥\" in user_message:\n",
    "        response = \"밥 먹으러 갑시다!\"\n",
    "    elif \"집\" in user_message:\n",
    "        response = \"집에 가즈아\"\n",
    "    print(response)\n",
    "\n",
    "    #응답 메세지 전송\n",
    "    await context.bot.send_message(chat_id=update.effective_chat.id, text=response)\n",
    "\n",
    "async def main():\n",
    "    application = Application.builder().token(TELEGRAM_TOKEN).build()\n",
    "    await application.initialize() # 초기화 필수\n",
    "\n",
    "    # 핸들러 추가\n",
    "    application.add_handler(CommandHandler(\"start\", start))\n",
    "    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_message))\n",
    "    \n",
    "    await application.start()\n",
    "    print(\"봇이 실행중 입니다.\")\n",
    "\n",
    "    await application.updater.start_polling()\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    asyncio.run(main())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73b7b49-cbc6-4439-804a-a87ecb51200d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
